compiler/haskell-ide-core/test/Demo.hs

```diff
@@ -0,0 +1,100 @@
+
+module Demo(main) where
+
+import Control.Concurrent.Extra
+import Control.Monad
+import System.Time.Extra
+import Development.IDE.State.Service
+import Development.IDE.State.Rules
+import Development.IDE.State.Shake
+import Development.IDE.State.RuleTypes
+import Development.IDE.Types.Diagnostics
+import Development.IDE.Types.Options
+import Development.IDE.Logger
+import qualified Data.Text.IO as T
+import System.Environment
+import Data.List
+import Data.Maybe
+import System.FilePath
+import Data.Tuple.Extra
+import System.IO.Extra
+import Development.IDE.Types.LSP
+import Development.Shake hiding (Env)
+import qualified Data.Set as Set
+
+import CmdLineParser
+import DynFlags
+import Panic
+import GHC
+import GHC.Paths
+
+
+main :: IO ()
+main = do
+    (ghcOptions, files) <- getCmdLine
+
+    -- lock to avoid overlapping output on stdout
+    lock <- newLock
+
+    ide <- initialise
+        mainRule
+        (Just $ showEvent lock)
+        (makeOneHandle $ withLock lock . T.putStrLn)
+        IdeOptions
+            {optPreprocessor = (,) []
+            ,optWriteIface = False
+            ,optGhcSession = liftIO $ newSession ghcOptions
+            ,optExtensions = ["hs"]
+            ,optPkgLocationOpts = error "optPkgLocationOpts not implemented yet"
+            ,optThreads = 0
+            ,optShakeProfiling = Nothing -- Just "output.html"
+            }
+    setFilesOfInterest ide $ Set.fromList files
+    _ <- runAction ide $ uses_ TypeCheck files
+    sleep 0.01 -- since Shake puts its completely message to stdout async
```

Comment:
What does this comment want to say?

Reply:
Revised, hopefully what it does now say

Reply:
Ah, it's about a "completed message". I didn't get that. Thanks for expanding.